src/VisualStudio/Core/Test/DocumentRefactoring/DocumentRefactoringServiceTests.vb

```diff
@@ -0,0 +1,107 @@
+
+Imports Microsoft.CodeAnalysis.Test.Utilities
+Imports Microsoft.CodeAnalysis.Editor.UnitTests.Workspaces
+Imports Microsoft.CodeAnalysis.Host
+Imports Microsoft.CodeAnalysis.LanguageServices
+Imports Microsoft.CodeAnalysis.Shared.Extensions
+Imports Microsoft.CodeAnalysis.Editor.UnitTests
+Imports Microsoft.VisualStudio.LanguageServices.Implementation.DocumentRefactoring
+Imports Microsoft.VisualStudio.Composition
+Imports System.IO
+Imports System.Xml
+
+Namespace Microsoft.VisualStudio.LanguageServices.UnitTests.DocumentRefactoring
+    <UseExportProvider>
+    Public Class DocumentRefactoringServiceTests
+        Private ReadOnly _catalog As ComposableCatalog = TestExportProvider.EntireAssemblyCatalogWithCSharpAndVisualBasic.WithPart(GetType(DocumentRefactoringService))
+        Private ReadOnly _factory As IExportProviderFactory = ExportProviderCache.GetOrCreateExportProviderFactory(_catalog)
+
+        Private Function TestDocumentRefactoring(markup As String, Optional expectMatchingName As Boolean = True) As Task
+            Dim workspace = TestWorkspace.CreateCSharp(markup, exportProvider:=_factory.CreateExportProvider())
+            Return TestDocumentRefactoring(workspace, expectMatchingName:=expectMatchingName)
+        End Function
+
+        Private Function TestDocumentRefactoring(files() As String, Optional ExpectMatchingName As Boolean = True) As Task
+            Dim workspace = TestWorkspace.CreateCSharp(files, exportProvider:=_factory.CreateExportProvider())
+            Return TestDocumentRefactoring(workspace, expectMatchingName:=ExpectMatchingName)
+        End Function
+
+        Private Async Function TestDocumentRefactoring(workspace As TestWorkspace, Optional expectMatchingName As Boolean = True) As Task
+            Dim project = workspace.CurrentSolution.Projects.Single()
+            Dim document = project.Documents.First()
+            Dim refactorService = workspace.GetService(Of DocumentRefactoringService)
+
+            Await refactorService.UpdateAfterInfoChangeAsync(document, document)
+
+            document = workspace.CurrentSolution.GetDocument(document.Id)
+
+            Dim syntaxRoot = Await document.GetSyntaxRootAsync().ConfigureAwait(False)
+
+            Dim syntaxFactsService = document.GetLanguageService(Of ISyntaxFactsService)
+            Dim typeDeclarationPairs = syntaxRoot _
+                                .DescendantNodes() _
+                                .Where(Function(n) syntaxFactsService.IsTypeDeclaration(n)) _
+                                .Select(Function(n) Tuple.Create(n, syntaxFactsService.GetDisplayName(n, DisplayNameOptions.None)))
+
+            Dim namespaces = syntaxRoot _
+                        .DescendantNodes()
+
+            Assert.True(typeDeclarationPairs.Any())
+            Dim matchingTypeDeclarationPair = typeDeclarationPairs.Where(Function(p) String.Equals(p.Item2, Path.GetFileNameWithoutExtension(document.FilePath), StringComparison.OrdinalIgnoreCase))
```

Comment:
Are we worried this test is just doing a copy/paste of the same code in the actual feature? Is it actually testing anything then?

Reply:
Would it be better to just do a markup test of before/after? Might make sense